Feature/java emitter enum

M1lney opened the pull request:
I think this is a proper implementation. From what I found the custom values in java need to be of the same type for each one.

M1lney committed:
Basic enum implementation.

M1lney committed:
Implement support for custom values for enum members #26

M1lney committed:
Testing combining enums #26

IlluzionzDev committed:
Merge branch 'alloy' into feature/java-emitter-enum

IlluzionzDev commented on packages/emitter-framework/src/java/components/enum.tsx:
```diff
@@ -0,0 +1,57 @@
+  let constructor = null;
+  let variable = null;
+  if (valueType) {
+    variable = <jv.Variable type={valueType} name="customValue" private={true}/>;
```
I would just rename the variable to be `value`, `customValue` seems too verbose

IlluzionzDev commented on packages/emitter-framework/src/java/components/enum.tsx:
```diff
@@ -0,0 +1,57 @@
+  if (valueType) {
+    variable = <jv.Variable type={valueType} name="customValue" private={true}/>;
+    constructor = <jv.Constructor name={type.name} parameters={{"customValue": valueType}}
+                                  children={"this.customValue = customValue;"} private={true}/>;
```
Few other feedback for this section:
- `private={true}` can just be `private`
- Passing `children` prop should be done through JSX not as a prop,
do something like this instead.
```
<jv.Constructor>
  this.value = value;
</jv.Constructor>
```
- Where you have `"customValue": valueType`, it can just be `customValue: valueType`

IlluzionzDev commented on packages/emitter-framework/src/java/components/enum.tsx:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  const joinedMembers = mapJoin(members, (member) => {
+    const customValue = member.value ? <jv.Value value={member.value}/> : null;
```
Where you have it returning `null` if there is no value, set it to `'null'` as if you are emitting the java `null`. At the moment it's not declaring a value for enums without a value and the java code isn't valid.

![image](https://github.com/user-attachments/assets/809d013b-28e5-485b-b3d5-bb24e21e748e)

IlluzionzDev replied:
Actually you can probably just do
```ts
const customValue = <jv.Value value={member.value} />;
```
As the value component will render `null` if member.value is undefined anyway

IlluzionzDev commented on packages/java-spring/src/emitter.tsx:
```diff
@@ -124,6 +124,15 @@ export async function $onEmit(context: EmitContext) {
             {emitResponseModels(serviceOperations)}
           </jv.PackageDirectory>
         </jv.PackageDirectory>
+        <jv.PackageDirectory package="enums">
```
Could you move these into the `models` directory aswell please

IlluzionzDev reviewed (changes requested):
Do you think you could write a quick unit test for this component aswell? Similar to the `model-declaration.test.tsx` one. If you don't have time don't worry.

M1lney committed:
Move generated enums to model package.
Pass null as value when one is not set in typespec
Remove any whitespace from enum members
clean up

M1lney committed:
Merge remote-tracking branch 'origin/feature/java-emitter-enum' into feature/java-emitter-enum

M1lney committed:
Move generated enums to model package.

M1lney committed:
Allow mixing members that have a value defined with those that do not.
Add some tests.

M1lney committed:
Remove code to remove white space as this is not needed

Add some tests

M1lney committed:
Provide get method for enum value

M1lney committed:
Update tests with get method changes
